### PR TITLE
Reports with projections

### PR DESCRIPTION
**Pull request by andrey-katz-systemorph**
The point of this PR is to allow projections in non interactive reports. Interactive reports will be dealt with separately.

### CODE REVIEW

File: ifrs17/Report/ReportStorage.ipynb [hunk 1]
```diff
@@ -110,6 +110,15 @@
                 "\n    protected readonly IExportVariable export;",
                 "\n    private readonly Systemorph.Vertex.Hierarchies.IHierarchicalDimensionCache hierarchicalDimensionCache;",
                 "\n    private readonly Systemorph.Vertex.Pivot.Builder.Interfaces.IPivotFactory reportFactory;",
+                "\n    private int periodicityInMonth;",
+                "\n",
+                "\n    public ProjectionConfiguration[] OrderedProjectionConfigurations {get; private set;}",
+                "\n",
+                "\n    public async Task<ProjectionConfiguration[]> GetOrderedProjectionConfigurationsAsync() => await workspace.Query<ProjectionConfiguration>()",
```
Comment (dcolleoni) [inline ifrs17/Report/ReportStorage.ipynb:117]: It's already here. But the GetOrderedProjectionConfigurationsAsync should react to changes of Args.Year and Args.Month. It is not a const.

File: ifrs17/Report/ReportStorage.ipynb [hunk 2]
```diff
@@ -132,11 +141,13 @@
                 "\n    // Constructor",
                 "\n    public ReportStorage(IWorkspace workspace, ",
                 "\n                        Systemorph.Vertex.Pivot.Builder.Interfaces.IPivotFactory reportFactory, ",
-                "\n                        IExportVariable export) {",
+                "\n                        IExportVariable export, ",
+                "\n                        int? periodicityInMonth = null) {",
```
Comment (dcolleoni) [inline ifrs17/Report/ReportStorage.ipynb:145]: this seems misplaced. It should be controlled like Year and Month. 
From a report you should be able to do: 
pv.Periodicity = 3.

Comment (dcolleoni) [reply ifrs17/Report/ReportStorage.ipynb:145]: It is part of the Args

File: ifrs17/Report/ReportMutableScopes.ipynb [hunk 3]
```diff
@@ -115,7 +115,9 @@
                 "\n    string Scenario { get; set; }  // TODO: Enable dropdown selection including All and Delta",
                 "\n    string Comparison { get; set; }  // TODO: only for scenario at the beginning, meant to enable general purpose comparisons ",
                 "\n    CurrencyType CurrencyType { get; set; }",
-                "\n    ",
+                "\n    int Projection {get; set;}",
+                "\n    protected ProjectionConfiguration[] projectionConfigurations => GetStorage().ProjectionConfigurations;",
```
Comment (dcolleoni) [inline ifrs17/Report/ReportMutableScopes.ipynb:119]: I think you can apply the FilterAndOrder(Month) here as this is a mutable scope and teh Month is a property which can be set from the outside.

File: ifrs17/Report/ReportConfigurationAndUtils.ipynb [hunk 4]
```diff
@@ -225,6 +228,20 @@
             "execution_count": 0,
             "outputs": []
         },
+        {
+            "cell_type": "code",
+            "source": [
+                "public static IDataCube<ReportVariable> SelectProjections(this IDataCube<ReportVariable> source, ",
```
Comment (dcolleoni) [inline ifrs17/Report/ReportConfigurationAndUtils.ipynb:234]: I would call it FilterProjections

File: ifrs17/Report/ReportConfigurationAndUtils.ipynb [hunk 5]
```diff
@@ -225,6 +228,20 @@
...
+                "\n                                                            int projectionLevel, ",
+                "\n                                                            ProjectionConfiguration[] orderedProjectionConfigurations) => source.Filter(x => {",
+                "\n    var relevantProjectionConfiguration = orderedProjectionConfigurations.Where((_, i) => i <= projectionLevel).Select(y => y.SystemName);",
```
Comment (dcolleoni) [inline ifrs17/Report/ReportConfigurationAndUtils.ipynb:237]: bring relevantProjectionConfiguration  outside the select as it doesn't need to be re-evaluated for each x

File: ifrs17/Constants/Consts.ipynb [hunk 6]
```diff
@@ -136,7 +136,16 @@
         {
             "cell_type": "code",
             "source": [
-                "public static int ProjectionCount = 1;"
+                "public static int ProjectionCount = 7 ;"
```
Comment (dcolleoni) [inline ifrs17/Constants/Consts.ipynb:139]: Reset this change

File: ifrs17/Utils/Queries.ipynb [hunk 7]
```diff
@@ -489,7 +489,11 @@
         {
             "cell_type": "code",
             "source": [
-                ""
+                "public static ProjectionConfiguration[] FilterAndOrder(this ProjectionConfiguration[] pc, int month) => ",
```
Comment (dcolleoni) [inline ifrs17/Utils/Queries.ipynb:492]: I would not have this in Queries NB. 
I could see it in Extensions.

Also rename it to SortRelevantProjections

File: ifrs17/Constants/Consts.ipynb [hunk 8]
```diff
@@ -136,7 +136,16 @@
...
+            ],
+            "metadata": {},
+            "execution_count": 0,
+            "outputs": []
+        },
+        {
+            "cell_type": "code",
+            "source": [
+                "public static int PeriodicityInMonth = 3;"
```
Comment (dcolleoni) [inline ifrs17/Constants/Consts.ipynb:148]: rename it MonthInAQuarter and move it at the end of section 2: Time Period

File: ifrs17/Report/ReportStorage.ipynb [hunk 9]
```diff
@@ -157,6 +162,7 @@
                 "\n    public async Task InitializeAsync((int year, int month) period, string reportingNode, string scenario, CurrencyType currencyType) {",
                 "\n        // Setting the Args --> Temp for the moment",
                 "\n        Args = (period, reportingNode, scenario, currencyType);",
+                "\n        ProjectionConfigurations = await GetProjectConfigurationsAsync();",
```
Comment (dcolleoni) [inline ifrs17/Report/ReportStorage.ipynb:165]: Write here the query and delete the method GetProjectConfigurationsAsync

File: ifrs17/Report/ReportStorage.ipynb [hunk 10]
```diff
@@ -173,7 +179,7 @@
                 "\n        foreach(var rn in GetLeaves<ReportingNode>(reportingNode)) {",
                 "\n            foreach(var scn in TargetScenarios) {",
                 "\n                if(!variablesDictionary.TryGetValue((period, rn, scn), out var variablesByIdentity)) {",
-                "\n                    variablesByIdentity = (await workspace.QueryReportVariablesAsync((period.year, period.month, rn, scn)))",
+                "\n                    variablesByIdentity = (await workspace.QueryReportVariablesAsync((period.year, period.month, rn, scn), ProjectionConfigurations.FilterAndOrder(Args.Period.Month)))",
```
Comment (dcolleoni) [inline ifrs17/Report/ReportStorage.ipynb:182]: use period.month instead of Args.Period.Month

**Comment by nnikolopoulos** [comment]
@andrey-katz-systemorph I noticed that this PR has as target branch `main` are you sure you want to merge to `main` and not `develop`?

**Comment by andrey-katz-systemorph** [comment]
> @andrey-katz-systemorph I noticed that this PR has as target branch `main` are you sure you want to merge to `main` and not `develop`?

Nice catch, thanks. I changed it.

**Review by dcolleoni — APPROVED** [review]

**Pull request merged**